### PR TITLE
Split QoD in Quality-On-Demand and QoS Profiles APIs

### LINKED ISSUE #265: Proposal to split QoS Sessions and QoS Profiles into two separate API definitions
**Problem description**

We have currently within the QoD API (v0.10.0) two different resources:

  - name: QoS Sessions
    description: Manage QoS sessions
  - name: QoS Profiles
    description: Manage QoS Profiles

There are several arguments why it would make sense to split the API in two separate ones:

* QoS Profiles could be used also in other use case (see comment below from @jlurien)
* QoS Sessions and QoS Profiles will (most) probably have different security and privacy requirements. Having them together within one API won't allow to use a "wildcard" scope across the complete API
* Operators who plan to offer only a small set of well-known QoS Profiles might want to decide to not implement QoS Profiles (at least initially). Currently that would lead to a partially implementation of the QoD API.
* The implementation of QoS Sessions and QoS Profiles might be based on different network and/or service components (e.g. service catalog), potentially implement by diffferent teams
* The current YAML is long and difficult to maintain (definitions for QoS Sessions and QoS Profiles are mixed up)

**Possible evolution**

Split the current API definition into two different
"QoS Sessions on Demand" (to be discussed to keep in the basepath `qod` for backward compatibility)
"QoS Profiles"

Potentially define a YAML with common definitions between the APIs (if not already available within https://github.com/camaraproject/Commonalities/blob/main/artifacts/CAMARA_common.yaml)

**Alternative solution**

?

**Additional context**

From discussion within #244:

> For applying scopes which allow the "wildcard" scope we might need also to think about splitting sessions and profiles in two APIs.

Splitting the API may have all sense. BTW, we are facing some new use cases where QoS profiles may be used by other APIs as well, so that would ease reusing this functionality.

_Originally posted by @jlurien in https://github.com/camaraproject/QualityOnDemand/issues/244#issuecomment-1932642055_

### PR DESCRIPTION
**Pull request by jlurien**
#### What type of PR is this?

* enhancement/feature

#### What this PR does / why we need it:

QoD v0.10.x API is split in 2:

- Quality-On-Demand, version WIP, with /sessions operations
  - PR suggests "Quality-On-Demand" and `/quality-on-demand`, as API Title and basePath 

- QoS Profiles, version WIP, with /qos-profiles operations
  - PR suggests "QoS Profiles" and `/qos-profiles`, as API Title and basePath 

The discussion about the names is not closed, so they may change.

Considerations:

- yaml files in /code folder. We may consider a different organization, e.g. a subfolder per API.
- All schemas moved to the spec where they are referred. Only the schema `QosProfileName` is used in both.
- securitySchemas is only kept in Quality-On-Demand. TBD which securityScheme to be applied to QoS Profiles, if any.
- Old `info.description` is kept mostly within Quality-On-Demand. A basic one is included for QoS Profiles, but may be enhanced. A cross reference between both APIs could be added.
- versions are "wip", and version in basePaths are `/v-wip`

#### Which issue(s) this PR fixes:

Fixes #265
Discussion is not closed. It was decided in last meeting that "we will discuss final names based on the PR." 

#### Special notes for reviewers:

README and other files in documentation will likely need to be adapted

#### Changelog input

```
- QoD API is split in 2: one for operations on sessions, and other for operations on qos-profiles

```

### CODE REVIEW

**Comment by jlurien** [comment]
This error requires some fixing in the CI, as now is dependent on the spec filename:

```
Run `npm audit` for details.
Run cd /home/runner/work/_actions/char0n/swagger-editor-validate/v1.3.2 && SWAGGER_EDITOR_URL=https://editor.swagger.io/ DEFINITION_FILE=code/API_definitions/qod-api.yaml IGNORE_ERROR= node src/index.js
Error: Error while validating in Swagger Editor
Error: Error: ENOENT: no such file or directory, open '/home/runner/work/QualityOnDemand/QualityOnDemand/code/API_definitions/qod-api.yaml'
Error: Process completed with exit code 1.
```

File: code/API_definitions/quality-on-demand.yaml [hunk 1]
```diff
@@ -719,146 +630,6 @@ components:
...
-          minimum: 1
-          maximum: 10
-          example: 3
-      required:
-        - name
-        - status
-
     QosProfileName:
       description: |
         A unique name for identifying a specific QoS profile.
```
Comment (sfnuser) [inline code/API_definitions/quality-on-demand.yaml:635]: I understand qos-profiles API is optional for quality-on-demand API. Would it be a good idea to refer to qos-profiles API? Something like `When qos-profile API is provided by the API provider, this QoSProfileName will be one from the profile set provided by the qos-profile API`

Or we could add this documentation to the description of this API in line 25, 45 or 51.

**Comment by hdamker** [comment]
From QoD meeting on May 3rd:
- Naming will be accepted as proposed for this PR ... can be discussed separately and changed if needed
- One week of review ... will be merged on next Friday (May 10th)  if no open comments.

**Comment by hdamker** [comment]
@camaraproject/quality-on-demand_maintainers - I just learned that a team will be removed from the reviewer list as soon as one member of the team has reviewed "on behalf" (here @RandyLevensalor). But I suppose that all team members have at least got the review request ones?

Anyway, this PR is important to get reviewed and merged, so that we continue will other issues. Please be aware of the deadline  of March 10th for your reviews.

**Comment by RandyLevensalor** [comment]
> This error requires some fixing in the CI, as now is dependent on the spec filename:
> 
> ```
> Run `npm audit` for details.
> Run cd /home/runner/work/_actions/char0n/swagger-editor-validate/v1.3.2 && SWAGGER_EDITOR_URL=https://editor.swagger.io/ DEFINITION_FILE=code/API_definitions/qod-api.yaml IGNORE_ERROR= node src/index.js
> Error: Error while validating in Swagger Editor
> Error: Error: ENOENT: no such file or directory, open '/home/runner/work/QualityOnDemand/QualityOnDemand/code/API_definitions/qod-api.yaml'
> Error: Process completed with exit code 1.
> ```

This fix has been merged.  When the branch is rebased it should resolve this issue.  @hdamker thanks for the quick fix.

File: code/API_definitions/quality-on-demand.yaml (hunk 1 above)
Comment (jlurien) [reply code/API_definitions/quality-on-demand.yaml:635]: Yes, I think it is convenient to document the relation with the other API, as now it is not self-contained in this API spec that other part.

Comment (jlurien) [reply code/API_definitions/quality-on-demand.yaml:635]: PR updated with a reference to the other API.

File: code/API_definitions/qos-profiles.yaml [hunk 2]
```diff
@@ -0,0 +1,475 @@
...
+    name: Apache 2.0
+    url: https://www.apache.org/licenses/LICENSE-2.0.html
+  version: wip
+externalDocs:
+  description: Product documentation at Camara
+  url: https://github.com/camaraproject/
+security:
+  - oAuth2ClientCredentials: []
+servers:
+  - url: "{apiRoot}/qos-profiles/v-wip"
```
Comment (RandyLevensalor) [inline code/API_definitions/qos-profiles.yaml:35]:
```suggestion
  - url: "{apiRoot}/qos-profiles/v-{version}"
```

Can we parameterize this?  Perhaps this needs to be a separate issue?

File: code/API_definitions/qos-profiles.yaml [hunk 3]
```diff
@@ -0,0 +1,475 @@
...
+    variables:
+      apiRoot:
+        default: http://localhost:9091
+        description: API root, defined by the service provider, e.g. `api.example.com` or `api.example.com/somepath`
+tags:
+  - name: QoS Profiles
+    description: Manage QoS Profiles
```
Comment (RandyLevensalor) [inline code/API_definitions/qos-profiles.yaml:42]:
```suggestion
    description: Query available QoS Profiles
```

We're only supporting get, post or delete calls that would modify the profile.  This is probably a holdover that was missed when we reduced the scope of the initial API.

**Review by RandyLevensalor — CHANGES_REQUESTED** [review]
Can the README.md also be updated to reflect these changes?

The links to the redoc and swagger views of these files will be missing along with some of the other descriptions.

**Comment by hdamker** [comment]
> Can the README.md also be updated to reflect these changes?
> 
> The links to the redoc and swagger views of these files will be missing along with some of the other descriptions.

@RandyLevensalor that will be done when we create a (public) release. Currently they are pointing correctly to the latest released version.

File: code/API_definitions/qos-profiles.yaml (hunk 2 above)
Comment (hdamker) [reply code/API_definitions/qos-profiles.yaml:35]: There is no simple 1:1 relation between "version" and the version within basepath depending on the type of API version. See https://wiki.camaraproject.org/display/CAM/API+versioning for details.

File: code/API_definitions/quality-on-demand.yaml [hunk 4]
```diff
@@ -73,16 +71,14 @@ externalDocs:
 security:
   - oAuth2ClientCredentials: []
 servers:
-  - url: "{apiRoot}/qod/v0"
+  - url: "{apiRoot}/quality-on-demand/v-wip"
```
Comment (hdamker) [inline code/API_definitions/quality-on-demand.yaml:74]: As above:

```suggestion
  - url: "{apiRoot}/quality-on-demand/vwip"
```

File: code/API_definitions/qos-profiles.yaml (hunk 2 above)
[the remark this answers is not in the document]
Comment (RandyLevensalor) [reply code/API_definitions/qos-profiles.yaml:35]: @hdamker cool.  I'd quit tracking the versioning discussion.  I'm happy to see the inclusion of beta and rc in version name.

[the remark this answers is not in the document]
Comment (hdamker) [reply code/API_definitions/qos-profiles.yaml:35]: @RandyLevensalor yes for alpha (internal use for reviews) and release candidates (to be used by implementations & to validate the definition) which are part of the release cycle. But there is not "beta", but instead the "initial" versions (0.y.z with the exception in the basepath as `/v0.y/`).

**Comment by hdamker** [comment]
Hi, as we have todo the QoD Call I kindly ask @camaraproject/quality-on-demand_maintainers for their review. Hopefully we can discuss today the outstanding comments and get the PR done.

**Review by sfnuser — APPROVED** [review]
Thanks.

File: code/API_definitions/qos-profiles.yaml [hunk 5]
```diff
@@ -0,0 +1,475 @@
...
+        "403":
+          $ref: "#/components/responses/Generic403"
+        "404":
+          $ref: "#/components/responses/QosProfileNotFound404"
+        "500":
+          $ref: "#/components/responses/QoSProfile500"
+        "503":
+          $ref: "#/components/responses/Generic503"
+
+components:
```
Comment (eric-murray) [inline code/API_definitions/qos-profiles.yaml:129]:
```suggestion
components:
  securitySchemes:
    oAuth2ClientCredentials:
      description: |
        The QoS Profiles API makes use of the OAUTH 2.0 client credentials grant which is applicable for server to server use cases involving trusted partners or clients without any protected user data involved. In this method the API invoker client is registered as a confidential client with an authorization grant type of client_credentials
      type: oauth2
      flows:
        clientCredentials:
          tokenUrl: https://api.example.com/oauth/token
          scopes: {}

```

File: code/API_definitions/qos-profiles.yaml [hunk 6]
```diff
@@ -0,0 +1,475 @@
...
+    email: project-email@sample.com
+  license:
+    name: Apache 2.0
+    url: https://www.apache.org/licenses/LICENSE-2.0.html
+  version: wip
+externalDocs:
+  description: Product documentation at Camara
+  url: https://github.com/camaraproject/
+security:
+  - oAuth2ClientCredentials: []
```
Comment (eric-murray) [inline code/API_definitions/qos-profiles.yaml:33]: Definition of `oAuth2ClientCredentials` is missing.

We could add the `openId` security scheme as part of these changes, but if keeping explicit OAuth2 client credentials for now, then it needs to be defined.

File: code/API_definitions/qos-profiles.yaml (hunk 5 above)
Comment (jlurien) [reply code/API_definitions/qos-profiles.yaml:129]: Agree. As the intention of the PR is just to split the current version, I think it's fine to keep by now the same definition. We will have to align all security related parts to the Guidelines before the meta-release.

File: code/API_definitions/qos-profiles.yaml (hunk 6 above)
Comment (jlurien) [reply code/API_definitions/qos-profiles.yaml:33]: As above, I prefer to keep it as it is right now, and leave the changes to align with guidelines for a later PR.

**Comment by jlurien** [comment]
@hdamker @eric-murray @RandyLevensalor Suggestions merged

**Review by hdamker — APPROVED** [review]

**Review by eric-murray — APPROVED** [review]

**Review by RandyLevensalor — APPROVED** [review]

**Pull request merged**